### ISSUE TITLE
..

### ISSUE THREAD

**Issue by Thanakool** [Thanakool/SimuLearn-Backend#1]
import pandas as pd
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as transforms
from PIL import Image

# ==========================================
# 1. การตั้งค่าและ Hyperparameters (ไม่ใช้ os)
# ==========================================
# ใช้ String ต่อกันตรงๆ (อิงจากโฟลเดอร์ปัจจุบัน)
DATA_DIR = "d1"  
IMAGE_DIR = DATA_DIR + "/img"
TRAIN_CSV = DATA_DIR + "/train.csv"
VAL_CSV = DATA_DIR + "/valid.csv"
TEST_CSV = DATA_DIR + "/test.csv"

# ตั้งค่าสำหรับเครื่องที่มี VRAM 20 GB
BATCH_SIZE = 64  
EPOCHS = 10
LEARNING_RATE = 1e-4
DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Mapping สำหรับคำถามและ Label Categorical
Q_COUNT = 'How many buildings are in this image?'
Q_DENSITY = 'What is the building density in this image?'
Q_AREA = 'What proportion of the image shows flood damage?'

DENSITY_MAP = {'no_building': 0, 'low': 1, 'moderate': 2, 'high': 3}
DENSITY_INV_MAP = {v: k for k, v in DENSITY_MAP.items()}

# ==========================================
# 2. การจัดการข้อมูล (Data Preprocessing)
# ==========================================
def pivot_dataframe(csv_path, is_test=False):
    df = pd.read_csv(csv_path)
    if is_test:
        return df[['filename']].drop_duplicates().reset_index(drop=True)
    
    q_map = {Q_COUNT: 'building_count', Q_DENSITY: 'density', Q_AREA: 'affected_area'}
    df['target'] = df['question'].map(q_map)
    df_pivot = df.pivot(index='filename', columns='target', values='answer').reset_index()
    
    df_pivot['building_count'] = df_pivot['building_count'].astype(float)
    df_pivot['affected_area'] = df_pivot['affected_area'].astype(float)
    df_pivot['density'] = df_pivot['density'].map(DENSITY_MAP)
    
    return df_pivot

class FloodDataset(Dataset):
    def __init__(self, df, img_dir, transform=None, is_test=False):
        self.df = df
        self.img_dir = img_dir
        self.transform = transform
        self.is_test = is_test

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        row = self.df.iloc[idx]
        img_name = row['filename']
        
        # ใช้เครื่องหมาย / ต่อ String ตรงๆ แทน os.path.join
        img_path = self.img_dir + "/" + img_name
        
        image = Image.open(img_path).convert("RGB")
        if self.transform:
            image = self.transform(image)

        if self.is_test:
            return image, img_name
        
        count = torch.tensor(row['building_count'], dtype=torch.float32)
        density = torch.tensor(row['density'], dtype=torch.long)
        area = torch.tensor(row['affected_area'], dtype=torch.float32)
        
        return image, count, density, area

# การทำ Data Augmentation และขยายขนาดภาพสำหรับ VRAM 20GB
train_transform = transforms.Compose([
    transforms.Resize((384, 384)),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
])

val_test_transform = transforms.Compose([
    transforms.Resize((384, 384)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
])

# ==========================================
# 3. การสร้างโมเดล (Custom CNN) ห้ามใช้ Pre-trained[span_1](start_span)[span_1](end_span)!
# ==========================================
class MultiTaskCNN(nn.Module):
    def __init__(self):
        super(MultiTaskCNN, self).__init__()
        self.features = nn.Sequential(
            nn.Conv2d(3, 32, kernel_size=3, padding=1), nn.BatchNorm2d(32), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(32, 64, kernel_size=3, padding=1), nn.BatchNorm2d(64), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(64, 128, kernel_size=3, padding=1), nn.BatchNorm2d(128), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(128, 256, kernel_size=3, padding=1), nn.BatchNorm2d(256), nn.ReLU(), nn.AdaptiveAvgPool2d((1, 1))
        )
        self.fc_shared = nn.Sequential(
            nn.Linear(256, 128), nn.ReLU(), nn.Dropout(0.3)
        )
        
        self.head_count = nn.Linear(128, 1)
        self.head_density = nn.Linear(128, 4)
        self.head_area = nn.Sequential(
            nn.Linear(128, 1), 
            nn.Sigmoid() 
        )

    def forward(self, x):
        x = self.features(x)
        x = torch.flatten(x, 1)
        x = self.fc_shared(x)
        
        out_count = self.head_count(x).squeeze(1)
        out_density = self.head_density(x)
        out_area = self.head_area(x).squeeze(1)
        
        return out_count, out_density, out_area

# ==========================================
# 4. ฟังก์ชันการ Train
# ==========================================
def train_model():
    print("Preparing Datasets...")
    df_train = pivot_dataframe(TRAIN_CSV)
    df_val = pivot_dataframe(VAL_CSV)
    
    train_dataset = FloodDataset(df_train, IMAGE_DIR, transform=train_transform)
    val_dataset = FloodDataset(df_val, IMAGE_DIR, transform=val_test_transform)
    
    train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=BATCH_SIZE, shuffle=False)
    
    model = MultiTaskCNN().to(DEVICE)
    optimizer = optim.Adam(model.parameters(), lr=LEARNING_RATE)
    
    criterion_count = nn.MSELoss()
    criterion_density = nn.CrossEntropyLoss()
    criterion_area = nn.MSELoss()
    
    print("Starting Training Loop...")
    for epoch in range(EPOCHS):
        model.train()
        train_loss = 0.0
        
        for images, counts, densities, areas in train_loader:
            images = images.to(DEVICE)
            counts = counts.to(DEVICE)
            densities = densities.to(DEVICE)
            areas = areas.to(DEVICE)
            
            optimizer.zero_grad()
            pred_counts, pred_densities, pred_areas = model(images)
            
            loss_c = criterion_count(pred_counts, counts)
            loss_d = criterion_density(pred_densities, densities)
            loss_a = criterion_area(pred_areas, areas)
            
            loss = loss_c + loss_d + loss_a
            loss.backward()
            optimizer.step()
            
            train_loss += loss.item()
            
        print(f"Epoch {epoch+1}/{EPOCHS} | Train Loss: {train_loss/len(train_loader):.4f}")

    torch.save(model.state_dict(), "model_day1.pth")
    print("Training Complete. Model saved as 'model_day1.pth'")
    return model

# ==========================================
# 5. ฟังก์ชันการ Inference และสร้างไฟล์ Submission
# ==========================================
def generate_submission(model):
    print("Generating Submission File...")
    df_test_unique = pivot_dataframe(TEST_CSV, is_test=True)
    test_dataset = FloodDataset(df_test_unique, IMAGE_DIR, transform=val_test_transform, is_test=True)
    test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False)
    
    model.eval()
    results = []
    
    with torch.no_grad():
        for images, img_names in test_loader:
            images = images.to(DEVICE)
            pred_counts, pred_densities, pred_areas = model(images)
            
            pred_counts = torch.round(pred_counts).clamp(min=0).cpu().numpy().astype(int)
            pred_densities = torch.argmax(pred_densities, dim=1).cpu().numpy()
            pred_areas = pred_areas.cpu().numpy()
            
            for i in range(len(img_names)):
                filename = img_names[i]
                results.append({'filename': filename, 'question': Q_COUNT, 'answer': str(pred_counts[i])})
                results.append({'filename': filename, 'question': Q_DENSITY, 'answer': DENSITY_INV_MAP[pred_densities[i]]})
                results.append({'filename': filename, 'question': Q_AREA, 'answer': str(pred_areas[i])})
                
    df_submission = pd.DataFrame(results)
    
    df_test_template = pd.read_csv(TEST_CSV)
    df_final = pd.merge(df_test_template[['filename', 'question']], df_submission, on=['filename', 'question'], how='left')
    
    df_final.to_csv("submission.csv", index=False)
    print("Submission saved to 'submission.csv'")

# ==========================================
# 6. รันโปรแกรมหลัก
# ==========================================
if __name__ == "__main__":
    trained_model = train_model()
    generate_submission(trained_model)

**Comment by Thanakool** [comment]
import pandas as pd
import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
import torchvision.transforms as transforms
from PIL import Image

# 1. ตั้งค่าพื้นฐาน
DATA_DIR = "d1"  
IMAGE_DIR = DATA_DIR + "/img"
TRAIN_CSV = DATA_DIR + "/train.csv"
VAL_CSV = DATA_DIR + "/valid.csv"
TEST_CSV = DATA_DIR + "/test.csv"

BATCH_SIZE = 64  # VRAM 20GB ดันขึ้น 64 หรือ 128 ได้เลย
EPOCHS = 10
LEARNING_RATE = 1e-4
DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")

Q_COUNT = 'How many buildings are in this image?'
Q_DENSITY = 'What is the building density in this image?'
Q_AREA = 'What proportion of the image shows flood damage?'

DENSITY_MAP = {'no_building': 0, 'low': 1, 'moderate': 2, 'high': 3}
DENSITY_INV_MAP = {v: k for k, v in DENSITY_MAP.items()}

# 2. จัดการ Data
def pivot_dataframe(csv_path, is_test=False):
    df = pd.read_csv(csv_path)
    if is_test:
        return df[['filename']].drop_duplicates().reset_index(drop=True)
    
    q_map = {Q_COUNT: 'building_count', Q_DENSITY: 'density', Q_AREA: 'affected_area'}
    df['target'] = df['question'].map(q_map)
    df_pivot = df.pivot(index='filename', columns='target', values='answer').reset_index()
    
    df_pivot['building_count'] = df_pivot['building_count'].astype(float)
    df_pivot['affected_area'] = df_pivot['affected_area'].astype(float)
    df_pivot['density'] = df_pivot['density'].map(DENSITY_MAP)
    
    return df_pivot

class FloodDataset(Dataset):
    def __init__(self, df, img_dir, transform=None, is_test=False):
        self.df = df
        self.img_dir = img_dir
        self.transform = transform
        self.is_test = is_test

    def __len__(self):
        return len(self.df)

    def __getitem__(self, idx):
        row = self.df.iloc[idx]
        img_name = row['filename']
        
        # ต่อ string ตรงๆ แทน os.path.join
        img_path = self.img_dir + "/" + img_name
        
        image = Image.open(img_path).convert("RGB")
        if self.transform:
            image = self.transform(image)

        if self.is_test:
            return image, img_name
        
        count = torch.tensor(row['building_count'], dtype=torch.float32)
        density = torch.tensor(row['density'], dtype=torch.long)
        area = torch.tensor(row['affected_area'], dtype=torch.float32)
        
        return image, count, density, area

train_transform = transforms.Compose([
    transforms.Resize((384, 384)),
    transforms.RandomHorizontalFlip(),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
])

val_test_transform = transforms.Compose([
    transforms.Resize((384, 384)),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
])

# 3. โมเดล (สร้างเองตามกฎเป๊ะๆ)
class MultiTaskCNN(nn.Module):
    def __init__(self):
        super(MultiTaskCNN, self).__init__()
        self.features = nn.Sequential(
            nn.Conv2d(3, 32, kernel_size=3, padding=1), nn.BatchNorm2d(32), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(32, 64, kernel_size=3, padding=1), nn.BatchNorm2d(64), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(64, 128, kernel_size=3, padding=1), nn.BatchNorm2d(128), nn.ReLU(), nn.MaxPool2d(2),
            nn.Conv2d(128, 256, kernel_size=3, padding=1), nn.BatchNorm2d(256), nn.ReLU(), nn.AdaptiveAvgPool2d((1, 1))
        )
        self.fc_shared = nn.Sequential(
            nn.Linear(256, 128), nn.ReLU(), nn.Dropout(0.3)
        )
        
        self.head_count = nn.Linear(128, 1)
        self.head_density = nn.Linear(128, 4)
        self.head_area = nn.Sequential(
            nn.Linear(128, 1), 
            nn.Sigmoid() 
        )

    def forward(self, x):
        x = self.features(x)
        x = torch.flatten(x, 1)
        x = self.fc_shared(x)
        
        out_count = self.head_count(x).squeeze(1)
        out_density = self.head_density(x)
        out_area = self.head_area(x).squeeze(1)
        
        return out_count, out_density, out_area

# 4. ลูป Train
def train_model():
    print("Preparing Datasets...")
    df_train = pivot_dataframe(TRAIN_CSV)
    df_val = pivot_dataframe(VAL_CSV)
    
    train_dataset = FloodDataset(df_train, IMAGE_DIR, transform=train_transform)
    val_dataset = FloodDataset(df_val, IMAGE_DIR, transform=val_test_transform)
    
    train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=BATCH_SIZE, shuffle=False)
    
    model = MultiTaskCNN().to(DEVICE)
    optimizer = optim.Adam(model.parameters(), lr=LEARNING_RATE)
    
    criterion_count = nn.MSELoss()
    criterion_density = nn.CrossEntropyLoss()
    criterion_area = nn.MSELoss()
    
    print("Starting Training Loop...")
    for epoch in range(EPOCHS):
        model.train()
        train_loss = 0.0
        
        for images, counts, densities, areas in train_loader:
            images = images.to(DEVICE)
            counts = counts.to(DEVICE)
            densities = densities.to(DEVICE)
            areas = areas.to(DEVICE)
            
            optimizer.zero_grad()
            pred_counts, pred_densities, pred_areas = model(images)
            
            loss_c = criterion_count(pred_counts, counts)
            loss_d = criterion_density(pred_densities, densities)
            loss_a = criterion_area(pred_areas, areas)
            
            loss = loss_c + loss_d + loss_a
            loss.backward()
            optimizer.step()
            
            train_loss += loss.item()
            
        print(f"Epoch {epoch+1}/{EPOCHS} | Train Loss: {train_loss/len(train_loader):.4f}")

    torch.save(model.state_dict(), "model_day1.pth")
    print("Training Complete. Model saved as 'model_day1.pth'")
    return model

# 5. สร้างไฟล์ Submit
def generate_submission(model):
    print("Generating Submission File...")
    df_test_unique = pivot_dataframe(TEST_CSV, is_test=True)
    test_dataset = FloodDataset(df_test_unique, IMAGE_DIR, transform=val_test_transform, is_test=True)
    test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE, shuffle=False)
    
    model.eval()
    results = []
    
    with torch.no_grad():
        for images, img_names in test_loader:
            images = images.to(DEVICE)
            pred_counts, pred_densities, pred_areas = model(images)
            
            pred_counts = torch.round(pred_counts).clamp(min=0).cpu().numpy().astype(int)
            pred_densities = torch.argmax(pred_densities, dim=1).cpu().numpy()
            pred_areas = pred_areas.cpu().numpy()
            
            for i in range(len(img_names)):
                filename = img_names[i]
                results.append({'filename': filename, 'question': Q_COUNT, 'answer': str(pred_counts[i])})
                results.append({'filename': filename, 'question': Q_DENSITY, 'answer': DENSITY_INV_MAP[pred_densities[i]]})
                results.append({'filename': filename, 'question': Q_AREA, 'answer': str(pred_areas[i])})
                
    df_submission = pd.DataFrame(results)
    
    df_test_template = pd.read_csv(TEST_CSV)
    df_final = pd.merge(df_test_template[['filename', 'question']], df_submission, on=['filename', 'question'], how='left')
    
    df_final.to_csv("submission.csv", index=False)
    print("Submission saved to 'submission.csv'")

if __name__ == "__main__":
    trained_model = train_model()
    generate_submission(trained_model)
